### ISSUE TITLE
Please make the wrapper a standalone tool

### ISSUE THREAD

**Issue by ximion** [gtkd-developers/GtkD#147]
Hi there!
I would like to use a library which provides GIR data with D, and the easiest way to do that at time is the wrapping tool found in GtkD.
It would be super-awesome to have the wrapper as a general-purpose tool to generate D code from any GIR file using just small metadata files.
I would definitely contribute to the new tool, and have already written a bit of code to make the wrapping too more generic - would be cool to make a generic gir-d-generator tool a reality.

What do you think?
Kind regards,
    Matthias

P.S: This is related to #24 , as a generic wrapping too might be able to generate missing D files on-demand.

**Comment by MikeWey** [comment]
That might be a good idea.
The current setup of parsing and generating everything each time probable isn't sustainable when more libraries are being generated.
Also the generation of Sourceview depends on Gtk and others, and these gir files and APILookup files have to be parsed to generate Sourceview correctly, so some sort of dependency tracking would be necessary. And would all the APILookup files live in the repository of the generator?

I would like to hear your thoughts on this.

**Comment by ximion** [comment]
> The current setup of parsing and generating everything each time probable isn't sustainable when more libraries are being generated.

Jup - ideally, D bindings would be generated as part of the library's own build process, which means that it would just call the `git-d-generator` and ship the resulting D bindings with the library. Alternatively, those bindings could be generated as part of another application's build process, or (worst option) be embedded.

> Also the generation of Sourceview depends on Gtk and others, and these gir files and APILookup files have to be parsed to generate Sourceview correctly, so some sort of dependency tracking would be necessary.

GIR files contain that information, fortunately. But we could likely also leave the dependency tracking up to Linux distributors, library vendors or simply `dub`.

> And would all the APILookup files live in the repository of the generator?

They would live in the library's source code or be provided individually be the person or project using the D binding generator.
Take a look at how Vala (a C#-inspired language that compiles to C) handles things with it's VAPI files (which are also generated from GIR):
https://wiki.gnome.org/Projects/Vala/Bindings 
https://wiki.gnome.org/Projects/Vala/UpstreamGuide
Using exactly the same approach for D would make a lot of sense, I think.

Also, the way linking is handled in the wrapper is really weird to me... I appears that the libraries are loaded dynamically, and the code then issues calls on the respective symbols (I might be wrong with that, as I am very new to D, still). I adjusted parts of the wrapper and resulting code to directly call the functions from D, and have `dub` link against the libraries I want.
Is there a reason for the linking behavior GtkD uses? Keep in mind I am a pure Linux guy and know very little about how Windows works.

In theory, it should be most efficient to directly link the D app/lib against a shared library. With my Debian developer hat on, this also makes packaging easier.
Would be cool if you could clear that up for me, since the GtkD approach might have advantages I don't see yet :-)

Thanks for your reply!

**Comment by ximion** [comment]
Hey :)
I published my current work here: https://github.com/ximion/gir-d-generator
The generator creates pretty well working libraries for all my testcases, and dub can link the required libraries perfectly well - with one exception: I need to link against `libgmodule2.0` which doesn't work due to this dub bug: https://github.com/D-Programming-Language/dub/issues/444

Things to do in future:
 1) Move some of the generic glib aliases (gchar, guint8, ...) to the generator itself - they are used everywhere, and you don't need to repeat the aliases everytime.
 2) Resolve dependencies of libraries - the GIR contains that information, ideally we should use it for the import line in D.
 3) More automagic: There are lots of things defined in the wrapper .txt files which you could automatically get from the GIR file.
 4) Nice CLI (e.g. define the gir name and output directory, and have the generator try to generate bindings and optionally also take a manually written manifest (.txt) file into account).
 5) Some build system integration (cmake/automake), so C/GObject projects can optionally generate D bindings upsream.

If you want to, we can move development of the gir-d-generator to some more neutral place, or even the GtkD umbrella. As said, this is incredibly useful, since lots of libraries ship GIR files which you now have access to from D.

**Comment by MikeWey** [comment]
Going forward i am thinking about the following:
1) Create a new repository under the gtkd-developers organisation, hopefully preserving the file history.
2) Restructure the wrapper as a standalone project.
3) Get it into a state where GtkD can use it to generate it's bindings.
4) future?

As for the linking behaviour, yes GtkD dynamically loads the libraries this way GtkD doesn't care about the version of GTK that is installed on your system, and if you have an older version the functions of the newer GTK versions are set to a default function that throws an exception.

For DMagick an D binding for ImageMagick i went the route of directly linking against the c libraries, and the translations of the c headers are filled with static if's to version out part of the API not present in older version.

Or you could say, GtkD 3.2 only works with GTK 3.18 and up.

**Comment by ximion** [comment]
> 1) Create a new repository under the gtkd-developers organisation, hopefully preserving the file history.

Preserving file history will be difficult, but if you want, I could do that for you - I already did this kind of split once with a different project and would just need to dig up the Git magic.
You would need to create the repo under the GtkD umbrella though, and if you want to give me commit access (otherwise I would just push this to a scratch repo and have you pull the code).

> 2) Restructure the wrapper as a standalone project.

That's something I've already done in the https://github.com/ximion/gir-d-generator repo - I use it here and it works well, although improvements should be made to make people write less verbose manifest files.

> 3) Get it into a state where GtkD can use it to generate it's bindings.

This should already work well - although I changed the generator to not dynamically load libraries, see below.

> As for the linking behaviour, yes GtkD dynamically loads the libraries this way GtkD doesn't care about the version of GTK that is installed on your system, and if you have an older version the functions of the newer GTK versions are set to a default function that throws an exception.

I figured that. My stance on this is that I want to have the project throw an error at build time - I don't want to ship it to a user and have them run into a bug when they use some functionality of the program which happens to rely on an unused symbol. Also, in the particular case I am working on, I also don't want to loose performance by dynamically loading libraries, although I didn't yet measure how big the impact of that would be.
The generated bindings should match the library I link against. For the gir-d-generator, this can easily be made possible by:
 * Having the upstream project generate the bindings in its build process - that way they're always up to date.
 * Generate the bindings dynamically at build time of the downstream project.
The latter option would likely be what people would choose. Since the GIR should always match the installed library version, there should be zero issues in that case, and we get proper build failures when the libraries change.

> For DMagick an D binding for ImageMagick i went the route of directly linking against the c libraries, and the translations of the c headers are filled with static if's to version out part of the API not present in older version.

That's a different case, since you want the manually written D binding to work with many libs - while for GIR based automatic binding generation, you just want to auto-generate a matching binding for the library you currently want to link against.
GObject Introspection is awesome, and one of the technologies developed at GNOME I like most :-) It makes generating bindings to many libraries for arbitrary programming languages real fun.
